classes/submission/SubmissionFileDAO.inc.php

```diff
@@ -781,6 +783,89 @@ public function getWorkflowStageId($submissionFile) {
+	 * @return array List of file stages (SUBMISSION_FILE_*)
+	 */
+	public function getAssignedFileStageIds($stageAssignments, $action) {
+		$allowedRoles = [ROLE_ID_MANAGER, ROLE_ID_SUB_EDITOR, ROLE_ID_ASSISTANT, ROLE_ID_AUTHOR];
+		$notAuthorRoles = array_diff($allowedRoles, [ROLE_ID_AUTHOR]);
+
+		$allowedFileStageIds = [];
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_SUBMISSION, $stageAssignments)
+				&& !empty(array_intersect($allowedRoles, $stageAssignments[WORKFLOW_STAGE_ID_SUBMISSION]))) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_SUBMISSION]));
+			// Authors only have read access
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_SUBMISSION;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_EDITING, $stageAssignments)
+				&& !empty(array_intersect($allowedRoles, $stageAssignments[WORKFLOW_STAGE_ID_EDITING]))) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_EDITING]));
+			// Authors only have read access
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_COPYEDIT;
+			}
+			if ($hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_FINAL;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_PRODUCTION, $stageAssignments)
+				&& !empty(array_intersect($allowedRoles, $stageAssignments[WORKFLOW_STAGE_ID_PRODUCTION]))) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_PRODUCTION]));
+			// Authors only have read access
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_PROOF;
+			}
+			if ($hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_PRODUCTION_READY;
+				$allowedFileStageIds[] = SUBMISSION_FILE_DEPENDENT;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_INTERNAL_REVIEW, $stageAssignments)
+				|| array_key_exists(WORKFLOW_STAGE_ID_EXTERNAL_REVIEW, $stageAssignments) ) {
+			$assignedUserRoles = array_merge(
+				$stageAssignments[WORKFLOW_STAGE_ID_INTERNAL_REVIEW] ?? [],
+				$stageAssignments[WORKFLOW_STAGE_ID_EXTERNAL_REVIEW] ?? []
+			);
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $assignedUserRoles));
+			// Authors can only write revision files under specific conditions
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_REVIEW_REVISION;
+			}
+			// Authors can never access review files
+			if ($hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_REVIEW_FILE;
+			}
+		}
+
+		HookRegistry::call('SubmissionFile::assignedFileStageIds', [&$allowedFileStageIds, $stageAssignments, $action]);
```

Comment:
Should the hook name be `SubmissionFileDAO`?

Reply:
Ah, yeah, hmm. So this is backporting from a method that exists in `SubmissionFileService` in the `master` branch. So it won't live in `SubmissionFileDAO` for long. Probably better to keep the hook name from changing, even if it's a little strange appearing in the `SubmissionFileDAO` for the stable branches. Does that sound ok?

Reply:
Yes, that's OK with me.